### PR TITLE
Refactor: isolate tasks and task comments as the separate components …

### LINKED ISSUE #738: Categorise mentorship_relation, tasks and comments
## Description
As a contributor,
I need better categorization of files and its related code logic,
so that I can easily find the related code and wrap up my head around the project.
### Currently the following are under one category
image

[resources/mentorship_relation.py](https://github.com/anitab-org/mentorship-backend/blob/develop/app/api/resources/mentorship_relation.py) and many other files can be further split into different files.
### We can categorise them as
- mentorship relation --> send, accept, delete, cancel, reject, current, past, pending
- tasks --> list, create, update, delete
- comments --> list, create, update, delete.

### Update [Required]
- [ ] Come with a solution to categorise them.
- [ ] Discuss all the required things we need to change and the existing code which might break after this change.

## Definition of Done
- [ ] All of the required items are completed.
- [x] Approval by 1 mentor.

## Estimation
[Can't estimate] hours

### PR DESCRIPTION
**Pull request by NenadPantelic**
### Description
Isolate endpoint related to tasks and task comments to the separate components (refactor DAO, DTO models and controllers). Fixes #738 and #740.

### Type of Change:

- Code
- Quality Assurance

**Code/Quality Assurance Only**
- Bug fix (non-breaking change which fixes an issue)

### How Has This Been Tested?
The existing tests were run. All of them pass (every endpoint that is refactored).
Screenshots:
image
image
image

### Checklist:
- [x] My PR follows the style guidelines of this project
- [x] I have performed a self-review of my own code or materials

**Code/Quality Assurance Only**

- [x] My changes generate no new warnings 
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/api/api_extension.py [hunk 1]
```diff
@@ -17,32 +26,33 @@ def ioslink():
     title="Mentorship System API",
     version="1.0",
     description="API documentation for the backend of Mentorship System. \n \n"
-    + "Mentorship System is an application that matches women in tech to mentor each other, on career development, "
+    + "Mentorship System is an application that matches women in tech to mentor each other, on career "
+    "development, "
     + "through 1:1 relations during a certain period of time. \n \n"
```
Comment (devkapilbansal) [inline app/api/api_extension.py]: Write it like this,
```suggestion
    "development, through 1:1 relations during a certain period of time. \n\n"
```

**Review by devkapilbansal — COMMENTED** [review]
Thanks @NenadPantelic 
There are a lot of changes which may take time in review.
Although the PR looks good

File: app/api/api_extension.py (hunk 1 above)
Comment (NenadPantelic) [reply app/api/api_extension.py]: Hi @devkapilbansal I think my Pycharm linter changed that. I used black formatter to reformat and apparently it missed that. Nevertheless, I'll change that. Thanks for the review :slightly_smiling_face:

**Comment by NenadPantelic** [comment]
@devkapilbansal I have changed the line breaking we discussed earlier. Can you re-review it? Thanks :slightly_smiling_face:

**Comment by NenadPantelic** [comment]
@vj-codes  @devkapilbansal  Any updates about this PR?

File: app/api/api_extension.py [hunk 2]
```diff
@@ -17,32 +26,32 @@ def ioslink():
     title="Mentorship System API",
     version="1.0",
     description="API documentation for the backend of Mentorship System. \n \n"
-    + "Mentorship System is an application that matches women in tech to mentor each other, on career development, "
-    + "through 1:1 relations during a certain period of time. \n \n"
-    + "The main repository of the Backend System can be found here: https://github.com/anitab-org/mentorship-backend"
+    + "Mentorship System is an application that matches women in tech to mentor each other, on career "
+    "development, through 1:1 relations during a certain period of time. \n \n"
```
Comment (epicadk) [inline app/api/api_extension.py:30]: why Is this change required ?

File: app/api/api_extension.py [hunk 3]
```diff
@@ -17,32 +26,32 @@ def ioslink():
...
-    + "\n\nFor more information about the project here's a link to our wiki guide: https://github.com/anitab-org/mentorship-backend/wiki"
-    + "\n\nThis <a href=https://github.com/anitab-org/mentorship-backend/blob/develop/docs/quality-assurance-test-cases.md>Quality Assurance Test cases</a> document contains examples of test scenarios to evaluate if the API is working as it should.",
+    + "\n\nGet started to using Backend Swagger UI here: "
+    "https://github.com/anitab-org/mentorship-backend/wiki/Using-Backend-Swagger-UI "
+    + "\n\nFor more information about the project here's a link to our wiki guide: "
+    "https://github.com/anitab-org/mentorship-backend/wiki "
+    + "\n\nThis <a href=https://github.com/anitab-org/mentorship-backend/blob/develop/docs/quality"
+    "-assurance-test-cases.md>Quality Assurance Test cases</a> document contains examples of test "
+    "scenarios to evaluate if the API is working as it should.",
 )
```
Comment (epicadk) [inline app/api/api_extension.py:46]: same here

Comment (NenadPantelic) [reply app/api/api_extension.py:46]: Hi @epicadk, by PEP8, lines were too long. Black approved this.

**Review by devkapilbansal — CHANGES_REQUESTED** [review]
@NenadPantelic just one minor change Replace all instances of strings with f-string in `app/api/resources/task.py`

**Comment by devkapilbansal** [comment]
Also, I would suggest to change the `Task Comment` to just `Comment`

**Comment by vj-codes** [comment]
> Also, I would suggest to change the `Task Comment` to just `Comment`

I say we keep it as task comments only also there is a file present in DAO in the same way and task comments make it more clear about the functionality and will be easier when implemented on the frontend too. What say?  @devkapilbansal  
And @NenadPantelic please solve the merge conflicts and  upload a screenshots of how the swagger ui looks now after refacotoring:)

**Comment by NenadPantelic** [comment]
> @NenadPantelic just one minor change Replace all instances of strings with f-string in `app/api/resources/task.py`

Hi @devkapilbansal, to be honest I just restructured code. I didn't change code style and things like string format. But sure, I'll do that. There are things also that should be changed in other code components, but that would be new issue.

**Comment by NenadPantelic** [comment]
> > Also, I would suggest to change the `Task Comment` to just `Comment`
> 
> I say we keep it as task comments only also there is a file present in DAO in the same way and task comments make it more clear about the functionality and will be easier when implemented on the frontend too. What say? @devkapilbansal
> And @NenadPantelic please solve the merge conflicts and upload a screenshots of how the swagger ui looks now after refacotoring:)

I agree with @vj-codes. I named it with the purpose to know those are task comments. In case we add some other comments feature. This way it is clear that these are comments that are related to the task. 
I will solve the merge conflict, they didn't exist at first. Probably some new merge made that.

**Comment by vj-codes** [comment]
> > @NenadPantelic just one minor change Replace all instances of strings with f-string in `app/api/resources/task.py`
> 
> Hi @devkapilbansal, to be honest I just restructured code. I didn't change code style and things like string format. But sure, I'll do that. There are things also that should be changed in other code components, but that would be new issue.

@devkapilbansal @NenadPantelic since the part in task.py was present in mentorship_relation.py before , PR #750 deals with changing those strings .
That PR is not merged and having other problems is a diff thing so the strings can either be ignored and create small issues later , or done in this PR leading to close of #750 .
Which way would you suggest?

**Comment by devkapilbansal** [comment]
@vj-codes that will make changes in mentorship-relation but here new files are created. That issue can still cover changes in that file. Also, as the repo is deleted the PR #750 should be closed and the issue should be made available again

**Comment by NenadPantelic** [comment]
Hi @devkapilbansal @vj-codes I made the changes you requested - recoded response messages to use f-style strings and added Swagger screenshots. Can you please review it again? :slightly_smiling_face:

**Review by devkapilbansal — DISMISSED** [review]
LGTM :+1: 

Thanks @NenadPantelic

**Review by vj-codes — CHANGES_REQUESTED** [review]
@NenadPantelic almost done!
Just one small change -
Please update the description for tasks to either
`Operations related to tasks for the mentee/mentor` OR
`Operations related to tasks for users in the current relation`

Because right now on android apk both mentee and mentor can create tasks , but for contributors going throughthe backend it will be misleading.Just this small grammatical change and then you're done!

**Comment by NenadPantelic** [comment]
> @NenadPantelic almost done!
> Just one small change -
> Please update the description for tasks to either
> `Operations related to tasks for the mentee/mentor` OR
> `Operations related to tasks for users in the current relation`
> 
> Because right now on android apk both mentee and mentor can create tasks , but for contributors going throughthe backend it will be misleading.Just this small grammatical change and then you're done!

@vj-codes Solved. I didn't squash this, I find it easy to review :slightly_smiling_face:

**Review by vj-codes — APPROVED** [review]
Thank you for your contribution @NenadPantelic 
LGTM 🎉

**Comment by vj-codes** [comment]
@devkapilbansal @epicadk can you review this once again , thanks!

**Review by devkapilbansal — APPROVED** [review]

**Review by codesankalp — APPROVED** [review]
The changes made in this PR was tested locally.

Following are the results:

1. Code review - Done

2. All possible responses were tested as mentioned below:
 * Task and Task Comments as separate component.
    *Screenshot/gif*:
image

Expected Result: Isolated task and task comments as separate components.
Actual Result: Task and task comments are actually separated components and isolated.

**Comment by codesankalp** [comment]
@isabelcosta @devkapilbansal @vj-codes this PR is tested locally.

**Review by isabelcosta — APPROVED** [review]
Cool refactor! @NenadPantelic 
Thank you for your contribution!
